### PR TITLE
DDG-107:  Added support for converting Mobi formatted files into EPub files.

### PR DESCRIPTION
**Pull request by sjacob**
@scoen @JonathanShaw 
This change adds support for converting Mobi formatted files into EPub files. It basically follows the pattern that the Jp2 and PDF converters use to convert files.

### CODE REVIEW

File: src/amberdb/model/Copy.java [hunk 1]
```diff
@@ -342,6 +346,79 @@ private void storeLegacyDossFile(File file, Path dossPath, String mimeType) thro
             file.putLegacyDoss(dossPath);
             file.setMimeType(mimeType);
         }
+        
+        @Override
+        public Copy deriveEPubCopy(Path epubConverterPath) throws Exception {
+            File file = this.getFile();
+            if (file == null) {
+                // Is not an image
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: Actually it means it has no file attached.

File: src/amberdb/model/Copy.java [hunk 2]
```diff
@@ -342,6 +346,79 @@ private void storeLegacyDossFile(File file, Path dossPath, String mimeType) thro
...
+                return null;
+            }
+            
+            // Do we need to check?
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: I don't know, do we?

Please either add a TODO or remove this comment.

File: src/amberdb/model/Copy.java [hunk 3]
```diff
@@ -342,6 +346,79 @@ private void storeLegacyDossFile(File file, Path dossPath, String mimeType) thro
...
+            String mimeType = file.getMimeType();
+            if (!(mimeType.matches("application/.*"))) {
+                throw new IllegalStateException(this.getWork().getObjId() + " is not a application/* type file. Unable to convert to EPub");
+            }
+
+            Path stage = null;
+            try {
+                // create a temporary file processing location for deriving the jpeg2000 from the master/comaster
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: I don't think you're making a jpeg2000?

File: src/amberdb/model/Copy.java [hunk 4]
```diff
@@ -342,6 +346,79 @@ private void storeLegacyDossFile(File file, Path dossPath, String mimeType) thro
...
+                        }
+                    }
+
+                    File acf = ac.getFile();
+                    acf.setFileSize(Files.size(epubPath));
+                    acf.setMimeType(epubMimeType);
+                }
+                
+                return ac;
+            } catch (Exception e) {
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: I'm guessing this is probably from our other methods, but please either do something with this, or remove it.

File: src/amberdb/model/Copy.java [hunk 5]
```diff
@@ -638,6 +715,47 @@ private Path generateImage(BlobStore doss, Path tiffUncompressor, Path jp2Genera
             if (!tiffUncompressor.toFile().exists() || !jp2Generator.toFile().exists()) return null;
             return jp2ImgPath;
         }
+        
+        private Path generateEPubFile(BlobStore doss, Path epubConverterPath, Path stage, Long blobId) throws Exception {
+            if (blobId == null) {
+                throw new NoSuchCopyException(this.getWork().getId(), CopyRole.fromString(this.getCopyRole()));
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: I'm not sure that this exception is appropriate. Maybe this should be NoSuchFile or something?

File: src/amberdb/model/Copy.java [hunk 6]
```diff
@@ -638,6 +715,47 @@ private Path generateImage(BlobStore doss, Path tiffUncompressor, Path jp2Genera
...
+            }
+
+            // prepare the files for conversion
+            Path tmpPath = stage.resolve("" + blobId);  // where to put the source retrieved from the amber blob
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: Please don't use <code>"" + blobId</code>

I realise this is probably copied from elsewhere, but nonetheless.

File: src/amberdb/model/Copy.java [hunk 7]
```diff
@@ -638,6 +715,47 @@ private Path generateImage(BlobStore doss, Path tiffUncompressor, Path jp2Genera
...
+                String mimetype = new Tika().detect(tmpPath.toFile());
+                switch (mimetype) {
+                    case "application/x-mobipocket-ebook": fileExtension = ".mobi"; break;
+                    case "application/vnd.amazon.ebook": fileExtension = ".azw"; break;
+                    default: throw new RuntimeException(mimetype + ": Not a file that can be converted to epub");
+                }
+            }
+
+            // Rename the file
+            String newFilename = "" + blobId + fileExtension;
```
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: as above.

**Comment by AaronCMuller** [comment]
Hi Simon,

Could you please negotiate with BSS that they are happy to install Calibre on all relevant servers before merging this request.

File: src/amberdb/util/EPubConverter.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+
+import java.nio.file.Path;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Convert an ebook into the EPub format using the Calibre ebook-convert command line tool.
```
Comment (brycollins) [inline src/amberdb/util/EPubConverter.java]: This seems more like a task for the edeposit servers to do before passing to banjo.
feels counter-productive to manage edeposit components across multiple systerms.

File: src/amberdb/model/Copy.java (hunk 5 above)
Comment (sjacob) [inline src/amberdb/model/Copy.java]: There isn't a NoSuchFileException defined in amberdb and I don't want to introduce one at this late stage of the project. NoSuchCopyException has been used for this purpose in other parts of the code.

File: src/amberdb/util/EPubConverter.java (hunk 8 above)
Comment (sjacob) [inline src/amberdb/util/EPubConverter.java]: I disagree. Banjo/Amberdb is the system responsible for generating derivative objects for repository items.

File: src/amberdb/model/Copy.java (hunk 5 above)
Comment (AaronCMuller) [inline src/amberdb/model/Copy.java]: Fair enough.

**Comment by sjacob** [comment]
@scoen @JonathanShaw 
I think this change is now ready to be merged.

**Pull request merged**